WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderSiteNotificationsUseCase.kt

```diff
@@ -0,0 +1,105 @@
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+/**
+ * This class handles reader notification events.
+ */
+class ReaderSiteNotificationsUseCase @Inject constructor(
+    private val dispatcher: Dispatcher,
+    private val analyticsUtilsWrapper: AnalyticsUtilsWrapper,
+    private val readerBlogTableWrapper: ReaderBlogTableWrapper,
+    private val networkUtilsWrapper: NetworkUtilsWrapper
+) {
+    private var continuation: Continuation<Boolean>? = null
+
+    suspend fun toggleNotification(blogId: Long): SnackbarMessageHolder? {
+        if (continuation != null) {
+            // Toggling notification for multiple sites in parallel isn't supported
+            // as the user would lose the ability to undo the action
+            return null
+        }
+        if (!networkUtilsWrapper.isNetworkAvailable()) {
+            return SnackbarMessageHolder(UiStringRes(R.string.error_network_connection))
+        }
+
+        // We want to track the action no matter the result
+        trackEvent(blogId)
+
+        val succeeded = suspendCoroutine<Boolean> { cont ->
+            continuation = cont
+            updateSubscription(blogId)
+        }
+
+        if (succeeded) {
+            updateBlogInDb(blogId)
+        }
+        return null
+    }
+
+    private fun trackEvent(blogId: Long) {
+        val trackingEvent = if (readerBlogTableWrapper.isNotificationsEnabled(blogId)) {
+            FOLLOWED_BLOG_NOTIFICATIONS_READER_MENU_OFF
+        } else {
+            FOLLOWED_BLOG_NOTIFICATIONS_READER_MENU_ON
+        }
+
+        analyticsUtilsWrapper.trackWithSiteId(trackingEvent, blogId)
+    }
+
+    private fun updateBlogInDb(blogId: Long) {
+        val isNotificationEnabledInDb = readerBlogTableWrapper.isNotificationsEnabled(blogId)
+        readerBlogTableWrapper.setNotificationsEnabledByBlogId(
+                blogId,
+                !isNotificationEnabledInDb
+        )
+    }
+
+    private fun updateSubscription(blogId: Long) {
+        val action = if (readerBlogTableWrapper.isNotificationsEnabled(blogId)) {
+            DELETE
+        } else {
+            NEW
+        }
+        val payload = AddOrDeleteSubscriptionPayload(blogId.toString(), action)
+        dispatcher.dispatch(AccountActionBuilder.newUpdateSubscriptionNotificationPostAction(payload))
+    }
+
+    @Subscribe(threadMode = ThreadMode.BACKGROUND)
+    @SuppressWarnings("unused")
+    fun onSubscriptionUpdated(event: OnSubscriptionUpdated) {
+        if (event.isError) {
+            continuation?.resume(false)
+            AppLog.e(
+                    API,
+                    ReaderSiteNotificationsUseCase::class.java.simpleName + ".onSubscriptionUpdated: " +
+                            event.error.type + " - " + event.error.message
+            )
+        } else {
+            continuation?.resume(true)
+            dispatcher.dispatch(AccountActionBuilder.newFetchSubscriptionsAction())
```

Comment:
NitPick: We might want to consider moving this call to `toggleNotification(..)` (line 56) so the whole flow is within that method. Wdyt?

Reply:
Moved here d7403d4, test updated in 44da620